components/raftstore/src/store/worker/read.rs

```diff
@@ -974,80 +974,155 @@ where
+                            &mut delegate,
+                            read_id,
+                            &mut snap_updated,
+                            last_valid_ts,
+                        ) {
+                            read_resp
+                        } else {
                             fail_point!("localreader_before_redirect", |_| {});
                             // Forward to raftstore.
                             self.redirect(RaftCommand::new(req, cb));
                             return;
                         }
-
-                        let region = Arc::clone(&delegate.region);
-                        let mut response =
-                            delegate.execute(&req, &region, None, Some(local_read_ctx));
-                        if let Some(snap) = response.snapshot.as_mut() {
-                            snap.bucket_meta = delegate.bucket_meta.clone();
-                        }
-                        // Try renew lease in advance
-                        delegate.maybe_renew_lease_advance(&self.router, snapshot_ts);
-                        response
                     }
                     // Replica can serve stale read if and only if its `safe_ts` >= `read_ts`
                     RequestPolicy::StaleRead => {
-                        let read_ts = decode_u64(&mut req.get_header().get_flag_data()).unwrap();
-                        if let Err(resp) = delegate.check_stale_read_safe(read_ts) {
-                            cb.set_result(ReadResponse {
-                                response: resp,
-                                snapshot: None,
-                                txn_extra_op: TxnExtraOp::Noop,
-                            });
-                            return;
+                        match self.try_local_stale_read(
+                            &req,
+                            &mut delegate,
+                            &mut snap_updated,
+                            last_valid_ts,
+                        ) {
+                            Ok(read_resp) => read_resp,
+                            Err(err_resp) => {
+                                // It's safe to change the header of the `RaftCmdRequest`, as it
+                                // would not affect the `SnapCtx` used in upper layer like.
+                                let unset_stale_flag = req.get_header().get_flags()
+                                    & (!WriteBatchFlags::STALE_READ.bits());
+                                req.mut_header().set_flags(unset_stale_flag);
+                                let mut inspector = Inspector {
+                                    delegate: &delegate,
+                                };
+                                // The read request could be handled using snapshot read if the
+                                // local peer is a valid leader.
+                                let allow_fallback_leader_read = inspector
+                                    .inspect(&req)
```

Comment:
Isn't this check little redundant? I agree with more checks and the metric might counted if the check failed here.

Reply:
The `inspector` function contains the whole read policy checks including `has_applied_to_current_term`. If the returned policy is `LocalRader` finally lease check would be done in the `try_local_leader_read` function.

Reply:
If `allow_fallback_leader_read` is false, it's also a failure of stale read fallback, what about increasing the `local_executed_stale_read_fallback_failure_requests` metric?

Reply:
If allow_fallback_leader_read is false, the peer is usually a follower peer which is not expected to be retried locally. Adding 1 to the counter for this situation seems to have little significance? Or do we prefer to record cases where we believe the leader can retry but the actual retry fails, indirectly indicating lease instability?

Reply:
> If `allow_fallback_leader_read` is false, the peer is usually a follower peer which is not expected to be retried locally.
I get it.